### PR TITLE
Post parsing/sync microservice

### LINKED ISSUE #1736: Create a "post parsing/sync" microservice
We are moving toward microservices for our back-end, and I want to refactor our current [Feed Parser](https://github.com/Seneca-CDOT/telescope/tree/master/src/backend/feed) into a separate service.

I don't know what to call this, some ideas:

- sync
- update
- parser

Whatever we call it, here is what it does:

- Connect the new User (#1642) and Posts (#1735) services via Redis
- Implements a Redis-backed job queue (we use https://github.com/OptimalBits/bull and there is also a new TypeScript port https://github.com/taskforcesh/bullmq that we could try)
- Manages getting feeds from the User service and creating jobs in the job queue
- Creates worker(s) to process jobs in the queue, parsing feeds and putting the data into Redis/Elasticsearch for use by the Posts service to consume.
- Keeps track of feeds that are invalid or causing errors, flags them in the User service
- Repeats the process of getting Feeds and Parsing them forever
- Provides an admin web dashboard of the job queue using Bull-Board https://github.com/vcapretz/bull-board

### PR DESCRIPTION
**Pull request by c3ho**
## Issue This PR Addresses
Fixes #1736 

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description
Still a WIP. Creates a post parsing microservice from our current code. There's a ton of files and I just want to get this up here for now. I opted to use the JS version of bull-queue, whether that is the correct choice or not I have no idea, but it shouldn't be too hard to make the switch to TS version in a future PR.

I started a small change in `processAllFeeds()` within the `index.js` file which will make a GET all user feeds request (that are not flagged) to the User ms.. I'll slowly improve this to do it in a pagination way.

I checked the Post api, there's no PUT routes yet to save all the Post data as per the most recent discussion. 

As satellite is waiting for Redis PR to be put in, this container will crash since I can't call `Redis()` yet. I'll also remove the test route I set up at `parserURL/`

Lists of tasks for this PR. I'll mark them off (maybe with a commit #) to track progress.
- [ ] Actually test this
- [ ] Write tests
- [x] Create readme
- [x] Create Dockerfile
- [ ] Update Feeds to store less info in Redis (need to doublecheck what we need and don't need)

## Checklist

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

**Review by manekenpix — CHANGES_REQUESTED** [review]
You probably need to rebase. `docker-compose-api-production.yml` doesn't exist anymore. Now it's `production.yml`.
Also, make sure you use the `env` files for port number or anything the microservice needs.

**Comment by humphd** [comment]
We need to consider the role of this service within the larger system we're building.  We have:

- User service (@chrispinkney in #1642): keeps track of users and their feeds
- Post service (@HyperTHD in #1735): deals with managing posts in Redis and Elastic, and has cross-over with this

This service should essentially manage the scheduling and processing of Feeds from to Posts.  A bunch of the code is already written, but it needs to be evolved a bit.  I think we need

- A Job Queue on top of Redis (we have this)
- A Job Worker process for extracting parsing jobs from the Job Queue, doing them, and sending results to the Posts service (we have most of this, but it needs to be taught to talk HTTP)
- A Job Scheduler, something that figures how when to go get Feeds and put them into the Job Queue to be processed again.  I'm not 100% clear on the best way to do this.  It could be time based, could be event based, ...

We used to use the Wiki as a source of truth, and that's being replaced by the User service.

@c3ho I see that you're using your favourite "Draft" style PR for this, but I think I'm going to need this to be done more collaboratively with the authors of the other services so that we can figure out our boundaries.  Probably this one is easiest to write last (after Users and Posts are merged).  That should happen for the 1.8 release in a few weeks.

Until then, maybe you can help review those other services with us as they come in, so we are all on the same page.

**Comment by HyperTHD** [comment]
Posts service has landed, so you can continue working on this service, and move out any Feed related stuff to your service, @c3ho

**Comment by c3ho** [comment]
@humphd @HyperTHD I don't think I should be managing any Post info here right?

How I think the service should work:
1. Send a request to User ms to get all feed urls.
2. Create queue and add feed urls to job list to process feed urls
3. Process feed urls and send the processed Post info to the Post ms to index in ElasticSearch/Redis 
4. Cache processed feeds in Redis
5. Rinse and repeat perpetually

In this case, I shouldn't care about how Post is implemented aside from sending parsed blog posts to the Post ms to store/index right? So in this case I am guessing I make a change to `articlesToPosts` function in `processor.js` to make
a POST request instead of what it  is doing currently?

**Comment by humphd** [comment]
@c3ho, I'm still not 100% sure of the proper design, but here are my thoughts:

The major thing this service does is maintain a job queue, and co-ordinate the creation of jobs.  How we do that is a bit of an open question.  We could poll, like we do now.  We could use [repeatable](https://optimalbits.github.io/bull/#repeatable) jobs (but we'd still have to maintain sync with the users/feeds in Firebase).  Somehow we have to get the feeds into this queue and process them, and repeat that.  So the easiest first method would be to get all the feeds from Firebase, yes.  It's probably not smart to "download all" like we do with the wiki, so maybe we iterate through the pages and add the jobs.  There might be a more "Firebase" way to do this, not sure.

Next we process the feeds like we do now, essentially, and when we're done, I think we put them directly into Redis/Elastic from this service, vs. involving the Posts service (it's read-only).

We need to keep Redis (cache) and Elastic (search) in sync with what's in Firebase.  Our feeds won't live in Redis anymore, we'll just cache posts there (with feed details that a post needs).  One way to do this would be to add one more key to Redis to track the current ordered set we're using to index posts: we can be showing the old set in the timeline while we build the new one, then delete the old one and continue the process.

We can iterate on this once the service exists.  If it mostly works like we have now, with User service vs. wiki feed list, it will be good.

**Comment by c3ho** [comment]
@humphd Thanks for the clarification. You're right, let's get the service working first then we'll reiterate / refactor.

**Comment by HyperTHD** [comment]
@c3ho Are you still interested in working on this PR? We're gonna need this in for 2.0 so someone else can take this issue if you're busy with something else

**Comment by c3ho** [comment]
I’ll have a reviewable PR for Friday

File: config/env.development
Comment (humphd) [inline config/env.development:129]: We have a lot of things running on 8***, let's pick something else, `10000`?

File: config/env.development
Comment (humphd) [inline config/env.development:132]: http vs. https

File: config/env.production
Comment (humphd) [inline config/env.production:158]: `https://api.telescope.cdot.systems/v1/parser`

File: config/env.staging
Comment (humphd) [inline config/env.staging:159]: `https://dev.api.telescope.cdot.systems/v1/parser`

File: docker/docker-compose.yml
Comment (humphd) [inline docker/docker-compose.yml:159]: I don't think you'll need posts.

File: docker/docker-compose.yml
Comment (humphd) [inline docker/docker-compose.yml:171]: Newline(s) after this

File: src/api/parser/jest.setup.js
Comment (humphd) [inline src/api/parser/jest.setup.js:2]: I think this has been done in the base config (cc @HyperTHD).

File: src/api/parser/package.json
Comment (humphd) [inline src/api/parser/package.json:20]: "^1.x"

File: src/api/parser/src/index.js
Comment (humphd) [inline src/api/parser/src/index.js:3]: What does this mean?

File: src/api/parser/src/index.js
Comment (humphd) [inline src/api/parser/src/index.js:15]: You're doing too much in this file.  You should split things up, so that this file just sets up the microservice with Satellite, and all this extra functionality can come from other modules.

File: src/api/parser/src/routes/parser.js
Comment (humphd) [inline src/api/parser/src/routes/parser.js:6]: This seems not worth doing.  Should we move what you have at `/queues` to this instead?  That is, going to `http://localhost/v1/parser` would show me the Bull Queues.

This also implies that we need some authentication/authorization (see Satellite, which has middleware you can use for this).  We either need to require admin access, or maybe just authenticated users can see it?  I'm not sure how sensitive this is.

**Review by humphd — CHANGES_REQUESTED** [review]

File: src/api/parser/src/routes/parser.js
Comment (c3ho) [reply src/api/parser/src/routes/parser.js:6]: Sure, we can disregard the whole `parser.js` file and just do `service.router.use('/', bullRouter)` in `index.js`

File: src/api/parser/src/index.js
Comment (c3ho) [reply src/api/parser/src/index.js:3]: This was supposed to have line 4 commented out as I wasn't sure if we'll need other routes aside from the bull-board

File: config/env.development
Comment (c3ho) [reply config/env.development:132]: All the other services in the file seem to be using http for local development

File: src/api/parser/jest.setup.js
Comment (HyperTHD) [reply src/api/parser/jest.setup.js:2]: Yup, this file is not needed anymore and can be removed

File: config/env.development
Comment (humphd) [reply config/env.development:132]: Yeah, it should be http.

**Comment by huynguyez** [comment]
@c3ho Any update on this PR?

**Comment by humphd** [comment]
@c3ho I anticipate this joining the critical path next week, once we get all the Users service work done this week.  If you want to keep going on it, cool. If not, let me know and we'll have someone take it over.

**Comment by c3ho** [comment]
@humphd sorry about the delay, I should be able to update the PR on it tomorrow.

**Comment by humphd** [comment]
@c3ho awesome, let me know when it's ready for another review.  Thanks.

File: config/env.development
Comment (humphd) [inline config/env.development:65]: You won't need to do this.  the `ALLOWED_APP_ORIGINS` is for the auth service to know about front-end origins that are OK to redirect a token back to.  You won't have any interaction with the front-end at all.

File: src/api/parser/package.json
Comment (humphd) [inline src/api/parser/package.json:7]: We should add some way to start the server here.  Get rid of `test` and add `start` and maybe `dev` if you want to do anything different there.

File: src/api/parser/readme.md
Comment (humphd) [inline src/api/parser/readme.md:15]: These don't exist, let's make sure this file represents what's really happening.

File: src/api/parser/readme.md
Comment (humphd) [inline src/api/parser/readme.md:1]: *cough*

File: src/api/parser/src/data/feed.js
Comment (humphd) [inline src/api/parser/src/data/feed.js:1]: You need to install this in your `package.json` here vs. in the root.

File: src/api/parser/src/feed/processor.js
Comment (humphd) [inline src/api/parser/src/feed/processor.js:7]: Missing dep

File: src/api/parser/src/feed/processor.js
Comment (humphd) [inline src/api/parser/src/feed/processor.js:8]: Missing dep

File: src/api/parser/src/feed/processor.js
Comment (humphd) [inline src/api/parser/src/feed/processor.js:10]: Move this up with the other 3rd party defines, and new line between them and local ones.

File: src/api/parser/src/lib/elastic.js
Comment (humphd) [inline src/api/parser/src/lib/elastic.js:2]: Missing dep

File: src/api/parser/src/lib/elastic.js
Comment (humphd) [inline src/api/parser/src/lib/elastic.js:3]: Missing dep

File: src/api/parser/src/lib/shutdown.js
Comment (humphd) [inline src/api/parser/src/lib/shutdown.js:1]: We might need to integrate this into how Satellite does its graceful shutdown code.  We need to investigate later.

File: src/api/parser/src/parser-queue.js
Comment (humphd) [inline src/api/parser/src/parser-queue.js:1]: fetch doesn't exist on satellite

File: src/api/parser/src/parser-queue.js
Comment (humphd) [inline src/api/parser/src/parser-queue.js:63]: For this to work we have to do a few things:

- In the docker-compose files, we have to declare that `USERS_URL` (note the trailing `S` on `USERS`) is part of the `environment` being passed into the process
- We need to specify the internal docker container domain for the localhost development case
- We need to create and pass a Service JWT Token to the Users Service so you can do this without being an authenticated user. 

You can checkout https://github.com/Seneca-CDOT/telescope/pull/2058/files, where I do similar things.

Comment (humphd) [reply src/api/parser/src/parser-queue.js:63]: Another thing we're going to need to do here is integrate a static JSON file with the current wiki feed list, so we can supplement the data (cc @manekenpix, @chrispinkney) with legacy feeds.  We probably need to do this in a follow-up PR vs. blocking this one on that, but FYI.

**Review by humphd — CHANGES_REQUESTED** [review]
We're going to have to think about how to write unit and e2e tests for this.

We might want to do that in follow-ups so we can get this in sooner.

File: config/env.development
Comment (c3ho) [reply config/env.development]: So sorry about this, I don't think this was supposed to be a change here.

File: src/api/parser/readme.md
Comment (humphd) [inline src/api/parser/readme.md:15]: You only have a `start` script above, better change this to `start`.

File: src/api/parser/src/data/feed.js
Comment (humphd) [inline src/api/parser/src/data/feed.js:113]: This stuff will probably need to be changed later to use what's in the Users data vs. what's in Redis.  For now, this is fine, but FYI so we don't forget.

File: src/api/parser/src/lib/queue.js
Comment (humphd) [inline src/api/parser/src/lib/queue.js:2]: This depends on @HyperTHD's PR in Satellite landing first.

**Review by humphd — CHANGES_REQUESTED** [review]

File: src/api/parser/package.json
Comment (humphd) [inline src/api/parser/package.json:5]: This can be removed.

File: src/api/parser/src/index.js
Comment (humphd) [inline src/api/parser/src/index.js:7]: Let's comment this out so that we can land on prod/staging today, and not have it compete with the existing backend that is also running a Bull queue in Redis.

**Review by humphd — CHANGES_REQUESTED** [review]

**Review by humphd — APPROVED** [review]

**Review by chrispinkney — APPROVED** [review]

**Pull request merged**
